### PR TITLE
Fix: bug ccg logic side effect on lexicon

### LINKED ISSUE #3345: ccg_semantics’ parser has undesirable side effects on the lexicon
Bug: when using nltk.ccg.chart with a lexicon that contains semantics, calling the parser may modify the lexicon object : namely, the semantics of the tokens may change during type raising and affect subsequent parses with this (modified) lexicon.

More precisely, type raising introduces new (bound) variables, which may appear as free variables in subsequent parses.

Minimal example:
```py
from nltk.ccg import chart, lexicon
from nltk.ccg.chart import printCCGDerivation

def lex(): return lexicon.fromstring(r"""
:- S, NP, N

V :: (S\NP)
Vt :: ((S\NP)/NP)
Det :: (NP/N)

cat => N {\x.cat(x)}
woman => N {\x. woman(x)}
sleeps => V {\x. sleep(x)}
pets => Vt {\x. \y. pet(y, x)}
the => Det {\P. the(P)}
of => ((N\N)/NP) {\e. \P. \x. (P(x) & belongs(x, e))}
that => ((N\N)/(S/NP)) {\S. \P. \x. (P(x) & S(x))}
""", True)

def sem_parser(): return chart.CCGChartParser(lex(), chart.DefaultRuleSet)


def parse_sentence(sentence, parser):
    for parse in parser().parse(sentence.split()):
        chart.printCCGDerivation(parse)
        break
    return


def bug():
    sentences = ["the cat of the woman sleeps", 
               "the cat that the woman pets sleeps"]
    print("WITH THE SAME LEXICON OBJECT")
    parser = sem_parser()
    for sentence in sentences:
        for parse in parser.parse(sentence.split()):
            chart.printCCGDerivation(parse)
            break
    print("Remark the free variable F in the semantics of the sentence")
    print("\n\n\n")
    print("WITH DIFFERENT LEXICON OBJECTS")
    for sentence in sentences:
        parse_sentence(sentence, sem_parser)
    print("Remark the correctness of the parsing")

bug()
```

With the following program, the semantics obtained for the second sentence is different depending on whether the first sentence was parsed before with the same lexicon.

In the example, the faulty semantics is `sleep(the(\x.(F(cat(x)) & pet(the(\z6.woman(z6)),x))))`, with the unwanted free variable `F` coming seemingly out of nowhere (it was introduced during type-raising in the previous parse)

### PR DESCRIPTION
**Pull request by ShadokDuBas**
See [issue 3345](https://github.com/nltk/nltk/issues/3345) for an in-detail explanation of the bug

Bug : In CCG, using a type-raising rule modifies the semantics of the words in the lexicon.
Fix : The function computing the semantics of a type-raising rules now makes a copy of its arguments, so as to not modify them.

### CODE REVIEW

**Comment by ekaf** [comment]
Thanks @ShadokDuBas, your PR yields the following semantics, which seem quite surprising:

>     S {sleep(the(\x.(cat(x) & belongs(x,the(\x.woman(x))))))} 

Is this a cat that is also a woman, and belongs to itself?

**Comment by arademaker** [comment]
The second `x` was introduced by the second quantification, under the scope of the first but the names should not collapse. In other words we could rename it to y without change the semantics (α-conversion). I am understanding the `the` over the lambda as existential quantification.

**Comment by ShadokDuBas** [comment]
You are right arademaker

I understand `the` as the function that takes a (non-empty) set and returns the contextually-salient element in it. The entity that sleeps is the salient entity that satisfies `λx.cat(x)` and `belong(x, the(λy.woman(y))`. It is strange that the quantification introduced by `the woman` uses the same variable is introduced by the other `the`, but it is mathematically correct (although confusing)… 

The only thing that would be problematic is that there is no way to reference to the cat inside the `woman` predicates. This would maybe cause an error if one wants the semantics of `the cat of the woman that feeds him sleeps` to be
```
sleep(the(\x.(cat(x) & belongs(x,the(\y.woman(y) & feeds(x, y) )))))
```
But setting the lexicon up so that `his` has a context dependent semantics is out of my abilities, so I cannot test, and I suspect that if it was done, then another variable name would be taken for `the woman that feeds him` (because x would appear as a free variable in the predicate already), and there would be no such bug (it happens because the semantics of woman is `λx. woman(x)`, but if the argument of `the` was more complex, some free variables would have been created)…

**Comment by ShadokDuBas** [comment]
For information note that it is the same result on the develop (original) branch, ie the parse
`sleep(the(\x.(cat(x) & belongs(x,the(\x.woman(x))))))` already appears, so my changes did not alter this (correct but confusing) result

**Comment by ekaf** [comment]
Yes @arademaker, the cat and the woman should not collapse in the first sentence. 

But @ShadokDuBas is right in pointing out that this bug is not introduced by this PR, since it also happens without the PR. It is also different from the bug that this PR intends to solve, namely the unwanted backpropagation of bindings from the chart into the lexicon object, which can lead to an imperfect coindexation of the cat in the second sentence, for ex. when a free variable has been introduced by parsing a previous sentence. We should not expect that this PR solves other bugs than this one, and it seems that _deepcopy_ is indeed the adequate means to block the backpropagation. 

On the other hand, the more we dig into this, the more problems seem to emerge.
CCG theory ensures the proper coindexation of semantic arguments in the relative clause through subscripted variables, and NLTK's ccg module provides the numbered series of "z" variables, which might be able to express the desired unifications, if an adequate variable binding mechanism is in place. The question is however whether the _ccg_ implementation is able to properly carry out those unifications. This would be easier to achieve in Prolog, where the underlying engine already provides all the correct bindings for free. So it seems possible that the semantic formulas in the _ccg_ module simply don't provide a sufficiently expressive notation to ensure adequate coindexations.

Last, there seems to be unsolved hurdles in CCG's treatment of relative phrases: in the [CCGbank manual](https://hockenmaier.cs.illinois.edu/Papers/CCGbank/CCGbankManual.pdf), there are several mentions (for ex. p. 28) that the CCG analysis of relative clauses may not always yield the desired semantic interpretation.

**Comment by ekaf** [comment]
@tanin47, could you advise whether _deepcopy_ is needed?

**Comment by ekaf** [comment]
Here is a draft by Gemini 1.5:

> This PR aims to address the issue of lexicon modification within the nltk.ccg.chart parser.
> However, after careful consideration and discussion, we believe that:
>  * Maintaining lexicon isolation is the primary use case for the nltk.ccg.chart parser. Reusing a chart should generally be avoided to prevent unintended lexicon modifications.
>  * The current behavior, while potentially surprising, is not necessarily a bug. If users are aware of the potential for lexicon modification, they can take appropriate steps to avoid it (e.g., creating a new chart for each parse).
>  * This PR might introduce unnecessary complexity and potential side effects while addressing a behavior that can be mitigated through proper usage and documentation.
> Recommendation:
>  * Reject PR #3349.
>  * Improve the documentation for nltk.ccg.chart:
>    * Clearly state that reusing a chart can modify the lexicon.
>    * Explain the implications of this behavior.
>    * Recommend creating new charts for each parse to maintain lexicon isolation.
> This approach will provide users with the necessary information to avoid unexpected behavior and maintain the desired level of lexicon isolation.
>

**Comment by ShadokDuBas** [comment]
If I understand correctly you present the point of view of Gemini. If I am wrong and you wrote this text, I ask you to forgive my harsh tone.
Having to disagree with randomly generated text was not on my wishlist for 2025, but here we go

> This PR aims to address the issue of lexicon modification within the nltk.ccg.chart parser.

True

> Maintaining lexicon isolation is the primary use case for the nltk.ccg.chart parser. Reusing a chart should generally be avoided to prevent unintended lexicon modifications.

I can agree… OR maybe it would be cool to not modify the lexicon

> The current behavior, while potentially surprising, is not necessarily a bug. If users are aware of the potential for lexicon modification, they can take appropriate steps to avoid it (e.g., creating a new chart for each parse).

"potentially surprising" ??????? If I create a lexicon A, then obtain a parser B, then use the parser B to parse a sentence C, no sane person would expect that the lexicon A is modified. If I create a greater framework to have semantics objects A, a pairing between these objects and words B, deduce from that a lexicon C, create a parser D, and parse a sentence with D, I would **NEVER** want my database A to be modified. This is a bug, not a feature.

> they can take appropriate steps to avoid it (e.g., creating a new chart for each parse).

We are talking about creating a new *lexicon*, not a new chart. What is more, what is the goal of creating charts that can parse multiple sentences if they cannot in fact parse multiple sentences ? The result of a parse is an enumerator, and yet every parse except the first one might be false because the first one messed with the data ?

> This PR might introduce unnecessary complexity 

True

> and potential side effects 

This PR’s goal is to resolve a side effect that is quite real and has proven to be a problem

> while addressing a behavior that can be mitigated through proper usage and documentation.

I don’t want to criticize too harshly the project, but saying that documentation is not its strong suit would be an understatement. I had to read the code to understand how things work.

> Clearly state that reusing a chart can modify the lexicon.

using, not reusing
   
> Explain the implications of this behavior.

"oh by the way, when you parse a sentence, yeah you know, you have to recompile everything, because parsing a sentence can break everything…" "It’s kind of like, you know, when you create a new bank account every time you want to receive your salary, because receiving a salary breaks the current bank account that you have"

> Recommend creating new charts for each parse to maintain lexicon isolation.

New *lexicons*, (ie rebuild everything from scratch) once again

> This approach will provide users with the necessary information to avoid unexpected behavior and maintain the desired level of lexicon isolation.

What about not relying on the user to work around the bugs but instead provide a function that already "avoid unexpected behavior" and "maintain the desired level of lexicon isolation" ? It is not that hard, it is literally one line of code ?



It is a bug, I can guarantee that whoever coded this in the first time did not want to modify the arguments. In fact this precise bug is very common in python because contrary to other languages, python implements passing by reference instead of passing by value. In other words, most languages copy the arguments of their functions, so that the function cannot modify it. Python does not, hence a function can modify the user-supplied data. Most of the time it is not wanted. The function copy.deepcopy is specifically used to not modify an argument : 
See for example [this question on stackoverflow](https://stackoverflow.com/questions/845110/emulating-pass-by-value-behaviour-in-python).

**Comment by ekaf** [comment]
Thanks @ShadokDuBas, I am discovering the ccg module, and don't know much about its implementation yet, but I expect that the truth will emerge through a healthy debate. 
The quoted text in my previous message was integrally written by Gemini. But I supplied the arguments during the conversation, through inputs like:

> But when we share the chart across different parses, the purpose must be to re-use earlier results, and then #3345 is not a bug, but an expected effect.

and

> 

> If maintaining a consistent and isolated state of the lexicon across parses is crucial, using deepcopy is not necessary, because then, the user should initialise a new chart instead of re-using a previous one.

This made Gemini evolve: initially it supported your point of view. I found it disappointing that Gemini did not contribute any original thinking to the discussion, but just repeated common sense phrases like a parrot. However I found that  it was good at expressing itself.
Now, I would like to hear more opinions about this: we need a thorough review of this issue, and my point of view can evolve also.

**Comment by ekaf** [comment]
@ShadokDuBas, after considering [your test](https://github.com/nltk/nltk/issues/3345#issuecomment-2558515950), I agree that instantiations made during parsing leave the lexicon object in a state that is most likely undesirable (maybe unless a user specifically wants to achieve this effect).

**Review by ekaf — APPROVED** [review]
This PR solves the unwanted backpropagation of free variables into the lexicon, described in #3345. I think that the ccg module needs this change and that this PR looks good to merge. 

However, identical output could be obtained by calling _compute_type_raised_semantics()_ with a deepcopy in _chart.py_, and leaving the _logic.py_ file unchanged. This is simpler, as it requires to change only one line. It seems also possible that a shallow copy (i.e. _copy.copy_) might be sufficient.

Other PRs will be needed for solving additional problems that came up: apparently, the _alpha_conversion_ is insufficient, and additional constraints might be needed, to prevent type raising from over-generating false analyses.

**Comment by stevenbird** [comment]
Sorry for the delay, and thanks to everyone for discussing the issues. This seems a critical shortcoming of the solution:
https://github.com/nltk/nltk/pull/3349/#issuecomment-2562391768
In other words, deep copy solves a problem but introduces another. I think it would preferable to document the issue than provide a partial solution.

**Comment by ekaf** [comment]
@stevenbird,  which "critical shortcoming of the solution" do you refer to? 

If it is the example where all the variables are collapsed to "x", it is not caused by this PR, because it also happens without the PR. It is due to to insufficient _alpha-conversion_, and is a different bug in the ccg module, solved in #3369. 

_Deepcopy_ has no unwanted side-effects here, but I noticed that _copy.copy_ also solves the bug. Still,  _deepcopy_ is safer, and it only induces additional overhead in the cases where it is needed. It is the same solution as the one included in #3369, which only  differs by a slightly different choice of variable names. 

So this PR looks good to merge before #3369, which is more comprehensive and will need additional review.

The CI failure has nothing to do with this PR, and was solved in #3372, so this branch just needs to incorporate the latest modifications of the _develop_ branch.

File: nltk/ccg/logic.py [hunk 1]
```diff
@@ -8,11 +8,14 @@
...
 """
 
+import copy
+
 from nltk.sem.logic import *
 
 
 def compute_type_raised_semantics(semantics):
-    core = semantics
+    semantics_copy = copy.deepcopy(semantics)
```
[the remark this answers is not in the document]
Comment (ekaf) [reply nltk/ccg/logic.py:17]: The review's suggestion to use copy instead of deepcopy is not recommendable, because there are only two possibilities:

- Either the structure is nested (contains other mutable objects inside), and then you need deepcopy to avoid side effects from shared references.
- If the structure is not nested (i.e., it only contains immutable or primitive types), deepcopy will not perform any expensive recursion, so the performance impact compared to copy is negligible.

Therefore, in either case, using deepcopy is the safe default: it only does the extra work when it is actually required, and otherwise behaves similarly to copy. This justifies using deepcopy in code where the structure might be nested or could change in the future.

**Comment by ekaf** [comment]
Defensive copying (`deepcopy`) is necessary in this context because neither leaving the CCG module unchanged nor pursuing alternatives is viable for real chart parsing:

- **Status quo (no copy):** The current approach mutates shared semantic objects in the lexicon during parsing, leading to unpredictable and non-reproducible results, subtle global-state bugs, and fundamentally undermining parser correctness and maintainability.

- **Alternatives:** Options like immutability, copy-on-write, or parser-level lexicon cloning are technically infeasible or disproportionately complex in a real chart parser context. They would require major architectural changes, risk subtle bugs, or introduce prohibitive performance and implementation overhead.

Defensive deep copying is a simple, robust solution that guarantees semantic isolation and parser correctness without a major refactor.

So, in a nutshell, there doesn't seem to exist a viable alternative to this PR.

**Comment by stevenbird** [comment]
Thanks @ekaf !
